[Skia] Building with USE_SKIA=ON still uses Cairo for ImageDiff

carlosgcampos opened the pull request:
#### 0e6f0245ab8d06d0bfa36cb19d3054d86b8ed798
<pre>
[Skia] Building with USE_SKIA=ON still uses Cairo for ImageDiff
<a href="https://bugs.webkit.org/show_bug.cgi?id=269190">https://bugs.webkit.org/show_bug.cgi?id=269190</a>

Reviewed by Alejandro G. Castro and Nikolas Zimmermann.

Add ImageDiff implementation for Skia.

* Source/ThirdParty/skia/CMakeLists.txt:
* Tools/ImageDiff/Cairo.cmake:
* Tools/ImageDiff/PlatformImage.h:
* Tools/ImageDiff/PlatformWPE.cmake:
* Tools/ImageDiff/Skia.cmake: Added.
* Tools/ImageDiff/skia/PlatformImageSkia.cpp: Added.
(ImageDiff::createFromStream):
(ImageDiff::PlatformImage::createFromFile):
(ImageDiff::PlatformImage::createFromStdin):
(ImageDiff::PlatformImage::createFromDiffData):
(ImageDiff::PlatformImage::PlatformImage):
(ImageDiff::PlatformImage::~PlatformImage):
(ImageDiff::PlatformImage::width const):
(ImageDiff::PlatformImage::height const):
(ImageDiff::PlatformImage::rowBytes const):
(ImageDiff::PlatformImage::hasAlpha const):
(ImageDiff::PlatformImage::pixels const):
(ImageDiff::PlatformImage::writeAsPNGToStdout):

Canonical link: <a href="https://commits.webkit.org/274544@main">https://commits.webkit.org/274544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fbc5d01255a92701eb1d10ca1b4827a07aa77fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39406 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41760 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41940 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35306 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41712 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21258 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15714 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32942 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39980 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/15481 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34132 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13447 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43218 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35775 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35404 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39214 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14218 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11721 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15824 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5163 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

alexgcastro reviewed (approved):
LGTM!

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+#include <skia/core/SkPixmap.h>
+#include <skia/core/SkStream.h>
+#include <skia/encode/SkPngEncoder.h>
+#include <stdio.h>
```
cstdio?

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "PlatformImage.h"
```
just asking: config.h include is not used in Tools/ImageDiff?

carlosgcampos replied:
No, ImageDiff doesn't have config.h, nor links to WTF or any other lib

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+    SkCodec::Result decodeResult;
+    auto codec = SkPngDecoder::Decode(std::move(stream), &decodeResult, nullptr);
+    if (decodeResult != SkCodec::Result::kSuccess) {
+        fprintf(stderr, "ImageDiff failed to decode PNG: %s\n", SkCodec::ResultToString(decodeResult));
```
We use fprintf here? No WTF logging?

carlosgcampos replied:
No WTF in ImageDiff

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+    auto imageInfo = SkImageInfo::Make(width, height, SkColorType::kGray_8_SkColorType, SkAlphaType::kOpaque_SkAlphaType);
+    SkPixmap pixmap(imageInfo, data, imageInfo.minRowBytes());
+    auto image = SkImages::RasterFromPixmap(pixmap, [](const void* pixels, void*) {
+        free(const_cast<void*>(pixels));
```
How can we be sure free() is sufficient? could be fastalloc'ed data that's passed here, no?

carlosgcampos replied:
There's not fast malloc in ImageDiff

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+{
+}
+
+PlatformImage::~PlatformImage()
```
= default?

nikolaszimmermann commented on Tools/ImageDiff/skia/PlatformImageSkia.cpp:
```diff
@@ -0,0 +1,136 @@
+
+PlatformImage::PlatformImage(sk_sp<SkImage>&& image)
+    : m_image(std::move(image))
+{
```
RELEASE_ASSERT(image), to make sure we can omit the nullpointer checks elsewhere?

carlosgcampos replied:
No WTF in Imagediff

nikolaszimmermann commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -891,6 +899,8 @@ target_include_directories(Skia INTERFACE
 target_compile_definitions(Skia PRIVATE
     SKIA_IMPLEMENTATION=1
 
+    SK_CODEC_DECODES_PNG
```
So we use the skia built-in png decoder for ImageDiff?

carlosgcampos replied:
Yes, ImageDiff only links to skia.

nikolaszimmermann reviewed (approved):
r=me, with a few comments - thanks @carlosgcampos!

carlosgcampos committed:
[Skia] Building with USE_SKIA=ON still uses Cairo for ImageDiff
https://bugs.webkit.org/show_bug.cgi?id=269190

Reviewed by Alejandro G. Castro and Nikolas Zimmermann.

Add ImageDiff implementation for Skia.

* Source/ThirdParty/skia/CMakeLists.txt:
* Tools/ImageDiff/Cairo.cmake:
* Tools/ImageDiff/PlatformImage.h:
* Tools/ImageDiff/PlatformWPE.cmake:
* Tools/ImageDiff/Skia.cmake: Added.
* Tools/ImageDiff/skia/PlatformImageSkia.cpp: Added.
(ImageDiff::createFromStream):
(ImageDiff::PlatformImage::createFromFile):
(ImageDiff::PlatformImage::createFromStdin):
(ImageDiff::PlatformImage::createFromDiffData):
(ImageDiff::PlatformImage::PlatformImage):
(ImageDiff::PlatformImage::~PlatformImage):
(ImageDiff::PlatformImage::width const):
(ImageDiff::PlatformImage::height const):
(ImageDiff::PlatformImage::rowBytes const):
(ImageDiff::PlatformImage::hasAlpha const):
(ImageDiff::PlatformImage::pixels const):
(ImageDiff::PlatformImage::writeAsPNGToStdout):

Canonical link: https://commits.webkit.org/274544@main
